feat: Add `two_step_if` function to `onebone.feature` module

Kangwhi-Kim opened the pull request:
Closes #41 

## Summary
_Write the summary about this PR._

- Add a function that estimates instantaneous frequency(IF) from a vibration signal.

## Changes
_Write changes in detail._

- A vibration-based IF estimation algorithm and its test code.
- For more information, refer to [the following paper](https://www.sciencedirect.com/science/article/abs/pii/S0888327012002270).

## How to Test?
_Write a way to test the changes._

- `python estimate_if.py`

## Checklist

- [x] Do you pass all the tests?
- [ ] Do you write the changes on document such as README?

Commit:
feat: Add a rpm tracking function and test code

- track_rpm.py is a function that estimate the rpm of
the vibration signal.
- test_track_rpm.py is a test code for track_rpm.py

Kangwhi-Kim commented:
## TODO LIST
- Add a docstring for `track_rpm`.

Kangwhi-Kim commented on onebone/feature/track_rpm.py:
```diff
@@ -0,0 +1,88 @@
+from typing import Union
```
### To do
파일 설명 및 만든이 인적사항 주석 필요

kbpark-onepredict commented on onebone/feature/track_rpm.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Make the size of frequency components equal to the size of the signal.
+    freq_components = interp1d(np.linspace(0, 1, t.size), freq_components)
+    freq_components = freq_components(np.linspace(0, 1, n))
```
freq_components라는 변수가 위쪽에서는 interpolation을 위한 함수의 역할을 하다가, 해당 줄에서 interpolation이 된 값으로 변화하는 것 같습니다.
동적 언어에서 충분히 가능한 구현 방법이긴 하지만, 가독성 측면에서 변수명을 나누어 역할을 명확히 해주는 것이 좋을 것 같습니다.

Kangwhi-Kim replied:
가독성 측면에서 좋은 의견입니다. 반영하겠습니다.

Commit:
feat: Changes the names of function

Changes name of track_rpm.py to tacholess.py.
Changes name of function `track_rpm` to `estimate_if`.

Commit:
feat: Changes for test case and Add the docstring

Added the docstring for estimate_if.py.

Kangwhi-Kim commented on onebone/feature/track_rpm.py:
```diff
@@ -0,0 +1,88 @@
+    noverlap: int = None,
+    **kwargs
+) -> np.ndarray:
+
```
반영했습니다.

Kangwhi-Kim commented on onebone/feature/track_rpm.py:
```diff
@@ -0,0 +1,88 @@
+    if not isinstance(x, np.ndarray):
+        raise TypeError("`x` must be array.")
+    if len(x.shape) >= 2:
+        raise ValueError("`x` has less than 2 dimensions.")
```
좋은 의견입니다. 반영하겠습니다.

Commit:
feat: Improve readability for code

Changes for variable names.

Commit:
feat: Modify typos

Modified typos for names of variables and documentation.

kyunghwan-onepredict commented on onebone/feature/tacholess.py:
```diff
@@ -0,0 +1,190 @@
+    return inst_freq
+
+
+def estimate_if(
```
저희가 메인으로 사용하는 함수들은 명사 형태로 작명하도록 했는데 다른 좋은 이름이 있을까요?
ex. `if_from_vib` ?

kyunghwan-onepredict replied:
`two_step_if` 라는 이름으로 논의하여 결정하였습니다.

kbpark-onepredict replied:
혹시 의미를 알려주실 수 있을까요? 직관적이진 않은 것 같습니다.

kyunghwan-onepredict commented on onebone/feature/tacholess.py:
```diff
@@ -0,0 +1,190 @@
+    # Get the size of the signal.
+    n = x.size
+    # Compute the Short Time Fourier Transform (STFT).
+    f, t, zxx = stft(x, fs, window, nperseg, noverlap, **kwargs)
```
`zxx`라는 변수를 scipy에서 사용하기는 한데 어떤 변수인지 이해하기가 어렵네요.
일반적으로 어떤 의미로 쓰는지? 다른 변수를 써도 된다면 더 좋은 변수로 하면 어떨까요

Kangwhi-Kim replied:
complex-value of stft coefficient인데, `coeff`로 변경했습니다.

kbpark-onepredict commented on onebone/feature/tacholess.py:
```diff
@@ -0,0 +1,190 @@
+
+    Parameters
+    ----------
+    x : numpy.ndarray of shape (length of `x`, )
```
해당 함수는 여러 채널의 센서에서 들어온 데이터가 들어갈 일이 없을까요?
있다면, 2차원도 지원 가능하게 변경해야하지 않을까요?

Kangwhi-Kim replied:
좋은 의견입니다. 하지만 `two_step_if`의 output이 clear하게 계산되기 위해서는 파라미터를 변경해가며 함수를 반복 사용해야합니다. 일반적으로, 어떤 하나의 signal에 대해서 `two_step_if`의 최적화된 파라미터가 있습니다. 따라서 여러 채널의 센서에서 각기 다른 signal을 '동일한' 파라미터로 연산하면 어느 signal의 output은 깔끔하게 계산되는 반면 어느 signal은 깔끔하지 않게 계산될 것입니다.
결과적으로, 여러 채널의 센서의 signal을 2차원화해서 `two_step_if` 사용하는 것보다 각 센서의 signal마다 각기 다른 파라미터의 `two_step_if`를 사용하는 것이 좋다고 생각합니다.

kbpark-onepredict replied:
감사합니다. 이해되었습니다.
다른 함수에서도 이러한 상황이 존재할 것 같습니다.
`moving_average`의 예시를 들면 smoothing의 목적이라면 신호의 특성마다 파라미터를 조절해주는게 좋을 것 같습니다.
생각보다 많은 수의 함수가 이러한 경우에 포함될 것 같은데 여기에 대한 기준이 필요하지 않을까요?
모든 함수는 1차원을 기준으로 작동하도록 하고, 2차원일 때 numpy.apply_along_axis로 래핑하는 방법을 통일한다던지

Kangwhi-Kim replied:
의견 감사합니다. 현재 기준은 '2차원으로 구현'하되, 2차원으로 구현하기 '어렵거나' 함수 사용면에서 '불필요'하면 1차원으로 구현하는 것으로 진행하고 있습니다. 예를 들어, `two_step_if`의 경우 2-d array의 모든 1-d array들을 동일한 함수 파라미터로 연산하는 경우가 희박하다고 판단되어 2차원으로 구현하지 않았습니다.

Kangwhi-Kim replied:
2차원 array 연산시 말씀하신 미리 구현된 decorator를 활용하는 방식은 좋다고 생각합니다. `onebone/util`에 넣으면 괜찮을 것 같네요.

Commit:
feat: Changes name of the function

Changes name of the function and variables

JunghoPark-OnePredict commented on onebone/feature/tacholess.py:
```diff
@@ -0,0 +1,190 @@
+        if f_range_indices[0].size == 0:
+            raise ValueError(
+                "Frequency cannot be tracked within a certain range `f_tol`. \
+                 Set `f_tol` or `nperseg` to a higher value than the current value."
```
해당 함수에는 'nperseg' 가 입력변수로 없는 것으로 보입니다.

JunghoPark-OnePredict commented on onebone/feature/tacholess.py:
```diff
@@ -0,0 +1,190 @@
+    return inst_freq
+
+
+def two_step_if(
```
함수 이름이 two_step_if 인데 instantaneous frequency 의 줄임말인 IF가 만약에의 영어인 if와 똑같아서 기능이 명확치 않을수도 있을것 같은데 어떻게 생각하시나요? matlab에서는 'instfreq' 로 표현하는것 같네요. 
https://www.mathworks.com/help/signal/ref/instfreq.html

Kangwhi-Kim replied:
의견 감사합니다. 저도 같은 의견으로 고민했습니다.
일단 IF(Instaneous frequency)는 여러 논문에서 자주 줄임말로 활용되고 있고, 함수 docstring에 자세히 설명하면 된다고 생각되어 `two_step_if`으로 정했습니다. (에디터에서 함수에 마우스를 가져다대면 설명이 나옵니다.) `onebone/feature`의 `mnf`(mean frequency), `mdf`(median frequency), `vcf`(variance of central frequency)도 같은 맥락입니다. 즉, 상충되지만 함수명을 최대한 간결하고 필요한 용어만 쓰되, 함수 docstring을 자세하게 서술하여 사용자에게 의미 전달을 하는데 집중했습니다.

JunghoPark-OnePredict replied:
의견 감사드립니다.
말씀하신 것을 들어보니 한가지 더 질문드릴게 생각나는데, 해당 함수를 signal에 넣는게 좋을지, feature에 넣는게 좋을지에 대해서는 어떻게 생각하시나요? 전체 신호에서 if를 계산하는 것이 signal processing 방법 같기도 해서 문의드립니다. feature는 좀 더 단일 scalar를 계산하는 함수들이 맞지도 않나 생각됩니다.

Kangwhi-Kim replied:
의견 감사합니다. 말씀해주신 덕분에 모듈 의미(성격)에 대한 정의를 다시 해보는 계기가 되었습니다.
다시 한번, 모듈에 구성된 함수들을 보면서 모듈 의미를 생각해보니,
### [`onebone/feature`](https://github.com/Onepredict/onebone/tree/main/onebone/feature) 모듈
1. 그 신호의 어떤 특징을 대표하는 값
2. signal --(추출)--> 1차 결과물  --(가공x)-->
3. 1차 결과물(특징)을 그대로 사용
4. 1차 결과물을 이어서 가공하지 않음
i.e. `mnf`, `mdf`, `vcf`, `na4`, `peak2peak`, `kurtosis`, `tacho_to_rpm` 등등
### [`onebone/signal`](https://github.com/Onepredict/onebone/tree/main/onebone/signal) 모듈 
1. 신호를 여러 목적에 따라 가공함
2. singal --(가공)--> 1차 결과물 --(가공)--> 2차 결과물 --> ...
3. 1차 결과물을 사용할 수 있지만 이어서 가공할 수 있음
i.e. `envelope_hilbert`, `postive_fft`, `lowpass_filter`, `moving_average` 등등

즉, 어떤 signal에 `onebone/signal` 모듈을 활용후 `onebone/feature` 모듈을 활용할 수 있지만, 역순으로 사용하지 않는 것 같습니다.

따라서, 위를 미루어 보았을 때 vibration signal에서 IF를 추출하는 함수는 `onebone/feature`에 가깝지 않을까 생각합니다. 하지만, 이것은 '어떤 관점으로 함수를 바라보냐' 에 따라 충분히 달라질 수 있다고 생각합니다. 피드백 주시면 감사합니다.

Kangwhi-Kim replied:
네, 오늘 회의 때 해당 내용에 대해 이야기해보려고 합니다. 제안해주셔서 감사합니다.

kyunghwan-onepredict reviewed (approved):
Great!

JunghoPark-OnePredict reviewed (approved):
수고 하셨습니다~!

Commit:
feat: Modify positive_fft example, merge origin/main into feat/track_rpm
